### ISSUE TITLE
turn post into a format 3 table?

### ISSUE THREAD

**Issue by madig** [googlefonts/googlesans#115]
I just saw that the `post` table takes up 28KB in the statics and VFs. Is there a hard need for it?

**Comment by m4rc1e** [comment]
We once changed `post.isFixedPitch` from proportional to mono in the Material Icons font and it did affect the fallback font in some browser (can't remember which one).  Even though this table was originally intended for postscript printers, I reckon some apps do use it for non-printing purposes.

**Comment by madig** [comment]
Hm, and what about the glyph names? I.e. turn `post` into a format 3 table? Those will occupy the most KBs.

**Comment by m4rc1e** [comment]
From Fontbakery:

```
      Apple recommends against using 'post' table format 3 under most           
      circumstances, as it can create problems with some printer drivers and    
      PDF documents. The savings in disk space usually does not justify the     
      potential loss in functionality.                                          
      Source: https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6post.html
```

**Comment by madig** [comment]
Oh well. Maybe if there are dedicated web fonts one day.

**Comment by moyogo** [comment]
A few bytes might be shaved off if glyph names are shortened to "u" + unicode, except for the standard Macintosh order glyph names.

**Comment by chrissimpkins** [comment]
https://typedrawers.com/discussion/3915/v2-vs-v3-post-table

**Comment by chrissimpkins** [comment]
How is this implemented?

**Comment by madig** [comment]
Hm? You set `font["post"].formatType = 3.0` and save.

**Comment by moyogo** [comment]
There's a ufo2ft lib key to generate TTFs with post format 3 but there's an issue with variable fonts: https://github.com/googlefonts/ufo2ft/issues/462.
Post processing as @madig suggest is the best way at the moment.

**Comment by anthrotype** [comment]
we still haven't figured out whether it is a issue. I think it's only an issue with the kind of font that Zachary is making

**Comment by chrissimpkins** [comment]
I am looking into it for file size reduction in another project where the downside is not as relevant as it is here. Ty!

**Comment by chrissimpkins** [comment]
Re the issue that Zachary posted and Denis mentioned above, this is the correct approach to do it with a lib key at the UFO source level?

```xml
<key>com.github.googlei18n.ufo2ft.keepGlyphNames</key> <false/>
```

**Comment by anthrotype** [comment]
Correct

**Comment by chrissimpkins** [comment]
I'm seeing a ~ 3% decrease in file size in a project with 126 glyphs using the ttLib.TTFont post-compile up-conversion approach.

> we still haven't figured out whether it is a issue.

I'll try the UFO lib key approach today and confirm that it works here.  I am confused by Zachary's thread.  My understanding is that the lib key approach does not work for fonts with massive (> 65k) glyph sets?

**Comment by anthrotype** [comment]
> My understanding is that the lib key approach does not work for fonts with massive (> 65k) glyph sets?

yes, that;s right. There seems to be an intrinsic limitation with the way post 2.0 is encoded

**Comment by anthrotype** [comment]
but it won't affect Google Sans

**Comment by chrissimpkins** [comment]
I think the issue for GS/GST is printer and pdf support.

**Comment by chrissimpkins** [comment]
I can confirm that the UFO lib key approach functions in this sub-65k glyph set build.  We have a post format 3 table that is binary identical to the ttLib.TTFont post-compile up-conversion approach.

**Comment by chrissimpkins** [comment]
Closing as we will not do this in this project, but this issue prompted me to do further investigation and we are using the approach for file size savings in derivatives that have targeted use env's where the post 3 table is not relevant.  Thanks for raising the issue Nikolaus.

**Comment by khaledhosny** [comment]
FIWIW, I checked fonts shipped with Windows a while ago and almost all of them used version 3 post table, so I think the printer driver issue is a red herring. The other issue I heard about post version 3 is that macOS autohinter might behave differently for certain glyphs if the font does not have names for them, but if this is is really an issue it probably concerns “standard” glyph names like AGL that someone might have hard-coded in the autohinter (Adobe’s autohinter have some hard-coded glyph names as well, but this works on source files so not an issue).

**Comment by khaledhosny** [comment]
I’m planning to use version 3 post table in Google Sans Arabic, as it would save ~13KB (in comparison, my GSUB table is ~5KB and GPOS is 35KB).

**Comment by chrissimpkins** [comment]
I reopened this so that we can discuss it Khaled.  I'm very open to it, but I suggest that we move the conversation over to the fontbakery repository.   post 3 is still a fail in fontbakery based on the information in [Apple documentation](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6post.html):

>Apple recommends against using 'post' table format 3 under most circumstances, as it can create problems with some printer drivers and PDF documents. The savings in disk space usually does not justify the potential loss in functionality.

FWIW, GS derivative projects that I prep for the Pixel Android environment have post 3 because it is irrelevant in that environment.  Apple continues to document that this could be an issue in some places.  Let's look into the legacy concerns and understand if they are relevant at all in the environments where GS lands.

**Comment by chrissimpkins** [comment]
https://github.com/googlefonts/fontbakery/issues/3635

**Comment by chrissimpkins** [comment]
It sounds like there is consensus that this is not a significant issue at this stage.  I'll submit a PR to downgrade the fontbakery check from FAIL to WARN.  @madig & @belluzj any reservations about bumping to post format 3 in Google Sans?

**Comment by madig** [comment]
Not from my side. Heh, we can then stop caring about PostScript names and our glyphset files will contain just glyph12345 and uniXXXX stuff :D

**Comment by chrissimpkins** [comment]
The update to downgrade this check was merged into fontbakery main.  Let's plan to do this in this upstream project too.  No urgency.

Ok to use post 3 downstream Khaled.  Thanks for pushing the issue again.

**Comment by chrissimpkins** [comment]
Based on Nikolaus's comment about grid fit differences in post 2 vs 3 fonts in https://github.com/googlefonts/fontbakery/issues/3635#issuecomment-1060765498, let's hold on doing this here with work that is already in production until I can sort out if this breaks the screenshot tests.  I wasn't expecting grid fit changes with post table format changes...

**Comment by chrissimpkins** [comment]
@madig What typeface family did you use in that example? (https://github.com/googlefonts/fontbakery/issues/3635#issuecomment-1060765498)

**Comment by madig** [comment]
That's our Aktiv Grotesk (Desktop vs. App font which basically is post v2 vs. v3)

**Comment by chrissimpkins** [comment]
related https://github.com/googlefonts/googlesans/issues/480

**Comment by madig** [comment]
Findings:

* A post table version 3.0 shows no diffs in diffenator3, and I didn't expect to see any.
* The shaping suite has to be completely regenerated, because suddenly most glyphs are named gid12345 instead of uni1234.ss01.
* A post table version 3.0 saves 98496 bytes in the upright VF.
